app/src/main/java/com/github/h3lp3rs/h3lp/forum/implementation/CachedForum.kt

```diff
@@ -0,0 +1,172 @@
+ * interface)
+ * @param forum The wrapped forum we want to cache
+ */
+class CachedForum(private val forum: Forum) : Forum {
+    override val path = forum.path
+    private val cache = storageOf(FORUM_CACHE)
+
+    // Entry for all the posts at one path
+    data class CacheEntry(val posts: ArrayList<ForumPostData>)
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    override fun newPost(
+        author: String,
+        content: String,
+        isPost: Boolean
+    ): CompletableFuture<ForumPost> {
+        // Cache the post if successfully returned
+        return forum.newPost(author, content, isPost).thenApply {
+            val post = ForumPost(CachedForum(it.forum), it.post, it.replies)
+            updateCacheWithPost(emptyList(), post.post)
+            post
+        }
+    }
+
+    override fun getPost(relativePath: Path): CompletableFuture<ForumPost> {
+        // Cache the post if successfully found
+        return forum.getPost(relativePath).thenApply {
+            val post = ForumPost(CachedForum(it.forum), it.post, it.replies)
+            updateCacheWithPost((path + relativePath).dropLast(1).drop(path.size), post.post)
+            post
+        }.exceptionally { exc -> // Fetch cache for post if not found
+            // Get the pointer of the category just above the required post
+            val cacheLoc = CachedForum(forum.child(relativePath.dropLast(1)))
+            // Get the post unique key
+            val key = relativePath.last()
+
+            val posts = cacheLoc.getCacheEntry(emptyList()).posts
+            var post: ForumPostData? = null
+            // Search for corresponding post in the category
+            for (p in posts) {
+                if (p.key == key) {
+                    post = p
+                }
+            }
+            post?.let {
+                val replies = cacheLoc.getCacheEntry(listOf(it.repliesKey)).posts
+                return@exceptionally ForumPost(cacheLoc, it, replies)
+            }
+
+            throw exc
+        }
+    }
+
+    override fun getAll(): CompletableFuture<List<CategoryPosts>> {
+        // Cache the posts if successfully returned
+        return forum.getAll().thenApply {
+            // Transform the list to make it recursively cache compatible
+            val cachedList = it.stream().map { l ->
+                CategoryPosts(l.first, l.second.stream().map { p ->
+                    ForumPost(CachedForum(p.forum), p.post, p.replies)
+                }.collect(toList()))
+            }.collect(toList())
+
+            for (categoryPosts in cachedList) {
+                val subForum = CachedForum(forum.root().child(categoryPosts.first))
+                for (post in categoryPosts.second) {
+                    subForum.updateCacheWithPost(emptyList(), post.post)
+                }
+            }
+            // Fetch cache for posts if not found -> If fail we only get an empty list
+            if (cachedList.isEmpty()) getAllAux() else cachedList

```

Comment:
Great comment! The future always succeeds even if it is offline! This is a design flaw in the SimpleDBForum which was too cumbersome to change. You can see the getAll function of SimpleDBForum and how it simply cannot fail. A bit sad and stupid but that's the reality (I spent a lot of time on trying to change this).

Reply:
To be more precise, the only thing that can return a failed future in the getAll function of SimpleDBForum is a failed getPost(). However, this does not propagate to the getAll() function properly, since an exception both means: no mapping and/or no connection. So the only way to fix this properly is to distinguish between both cases, way to hard and cumbersome for an already long 7h task.